File: 2021/tests/exercises/test_exercise_12.py
from adventofcode.exercises import exercise_12
from adventofcode import utils

TEST_DATA_SMALL = """start-A
start-b
A-c
A-b
b-d
A-end
b-end"""

TEST_DATA_LARGE = """dc-end
HN-start
start-kj
dc-start
dc-HN
LN-dc
HN-end
kj-sa
kj-HN
kj-dc
"""

TEST_DATA_EVEN_LARGER = """fs-end
he-DX
fs-he
start-DX
pj-DX
end-zg
zg-sl
zg-pj
pj-he
RW-he
fs-DX
pj-RW
zg-RW
start-pj
he-WI
zg-he
pj-fs
start-RW"""


def test_small_example_on_exercise_a():
    assert exercise_12.run_a(TEST_DATA_SMALL) == 10


def test_large_example_on_exercise_a():
    assert exercise_12.run_a(TEST_DATA_LARGE) == 19


def test_even_larger_example_on_exercise_a():
    assert exercise_12.run_a(TEST_DATA_EVEN_LARGER) == 226


def test_returns_correct_answer_on_exercise_a():
    input_data = utils.load_data(exercise_12.EXERCISE)
    assert exercise_12.run_a(input_data) == 3463


def test_small_example_on_exercise_b():
    assert exercise_12.run_b(TEST_DATA_SMALL) == 36


def test_large_example_on_exercise_b():
    assert exercise_12.run_b(TEST_DATA_LARGE) == 103


def test_even_larger_example_on_exercise_b():
    assert exercise_12.run_b(TEST_DATA_EVEN_LARGER) == 3509


def test_returns_correct_answer_on_exercise_b():
    input_data = utils.load_data(exercise_12.EXERCISE)
    assert exercise_12.run_b(input_data) == 91533

File: 2024/adventofcode/tests/test_exercise_12.rs
use adventofcode::exercises::exercise_12;
use adventofcode::exercises::utils;

const TEST_DATA_SMALL: &str = "AAAA
BBCD
BBCC
EEEC";

const TEST_DATA_MEDIUM: &str = "OOOOO
OXOXO
OOOOO
OXOXO
OOOOO";

const TEST_DATA_LARGE: &str = "RRRRIICCFF
RRRRIICCCF
VVRRRCCFFF
VVRCCCJFFF
VVVVCJJCFE
VVIVCCJJEE
VVIIICJJEE
MIIIIIJJEE
MIIISIJEEE
MMMISSJEEE";

const TEST_DATA_MEDIUM_2: &str = "EEEEE
EXXXX
EEEEE
EXXXX
EEEEE";

const TEST_DATA_MEDIUM_3: &str = "AAAAAA
AAABBA
AAABBA
ABBAAA
ABBAAA
AAAAAA";

#[test]
fn test_a_returns_correct_answer_for_test_data_small() {
    assert_eq!(exercise_12::run_a(TEST_DATA_SMALL), 140);
}

#[test]
fn test_a_returns_correct_answer_for_test_data_medium() {
    assert_eq!(exercise_12::run_a(TEST_DATA_MEDIUM), 772);
}

#[test]
fn test_a_returns_correct_answer_for_test_data_large() {
    assert_eq!(exercise_12::run_a(TEST_DATA_LARGE), 1930);
}

#[test]
fn test_a_returns_correct_answer_on_exercise_a() {
    let input_data = utils::load_data(exercise_12::EXERCISE);
    assert_eq!(exercise_12::run_a(&input_data), 1431316);
}

#[test]
fn test_b_returns_correct_answer_for_test_data_small() {
    assert_eq!(exercise_12::run_b(TEST_DATA_SMALL), 80);
}

#[test]
fn test_b_returns_correct_answer_for_test_data_medium() {
    assert_eq!(exercise_12::run_b(TEST_DATA_MEDIUM), 436);
}

#[test]
fn test_b_returns_correct_answer_for_test_data_medium_2() {
    assert_eq!(exercise_12::run_b(TEST_DATA_MEDIUM_2), 236);
}

#[test]
fn test_b_returns_correct_answer_for_test_data_medium_3() {
    assert_eq!(exercise_12::run_b(TEST_DATA_MEDIUM_3), 368);
}
#[test]
fn test_b_returns_correct_answer_for_test_data_large() {
    assert_eq!(exercise_12::run_b(TEST_DATA_LARGE), 1206);
}

#[test]
fn test_b_returns_correct_answer_on_exercise_b() {
    let input_data = utils::load_data(exercise_12::EXERCISE);
    assert_eq!(exercise_12::run_b(&input_data), 821428);
}
